runtime/logging/logger.go

```diff
@@ -38,100 +38,105 @@ type Options struct {
+}
 
-	// We add one to frameskip because we also skip makeEntry's frame.
-	_, file, line, ok := runtime.Caller(frameskip + 1)
-	if ok {
-		entry.File = file
-		entry.Line = int32(line)
-	}
+var _ slog.Handler = &LogHandler{}
 
-	return &entry
+// Handle implements the slog.Handler interface.
+func (h *LogHandler) Handle(rec slog.Record) error {
+	h.Write(h.makeEntry(rec))
+	return nil
 }
 
-// FuncLogger is a logger that calls a supplied function on every log entry.
-type FuncLogger struct {
-	Opts  Options                      // configures the log entries
-	Write func(entry *protos.LogEntry) // called on every log entry
+// Enabled implements the slog.Handler interface.
+func (h *LogHandler) Enabled(context.Context, slog.Level) bool {
+	// Support all logging levels.
+	return true
 }
 
-var _ logtype.Logger = FuncLogger{}
-
-// Debug implements the [weaver.Logger] interface.
-func (l FuncLogger) Debug(msg string, attrs ...any) {
-	l.Write(makeEntry("debug", msg, attrs, 1, l.Opts))
+// WithAttrs implements the slog.Handler interface.
+func (h *LogHandler) WithAttrs(attrs []slog.Attr) slog.Handler {
+	// Note that attributes set through explicit calls to log.WithAttrs should
+	// apply to all the log entries, similar to h.Opts.
+	//
+	// Note that WithAttrs results in a new logger, hence we should create a new
+	// handler that contains the new attributes.
+	rh := &LogHandler{
+		Opts:  h.Opts,
+		Write: h.Write,
+	}
+	rh.Opts.Attrs = appendAttrs(rh.Opts.Attrs, attrs)
+	return rh
 }
 
-// Info implements the [weaver.Logger] interface.
-func (l FuncLogger) Info(msg string, attrs ...any) {
-	l.Write(makeEntry("info", msg, attrs, 1, l.Opts))
+// WithGroup implements the slog.Handler interface.
+//
+// TODO(rgrandl): Implement it, so the users have the same experience as slog
+// if they decide to use WithGroup.
+func (h *LogHandler) WithGroup(string) slog.Handler {
+	return h
 }
 
-// Error implements the [weaver.Logger] interface.
-func (l FuncLogger) Error(msg string, err error, attrs ...any) {
-	e := makeEntry("error", msg, attrs, 1, l.Opts)
-	if err != nil {
-		e.Attrs = append(e.Attrs, "err", err.Error())
+// makeEntry returns an entry that is fully populated with information captured
+// by a log record.
+func (h *LogHandler) makeEntry(rec slog.Record) *protos.LogEntry {
+	// TODO(sanjay): Is it necessary to copy opts.Attrs even if no new attrs
+	// are being added?
+	var attrs []slog.Attr
+	rec.Attrs(func(a slog.Attr) {
+		attrs = append(attrs, a)
+	})
+
+	entry := protos.LogEntry{
+		App:        h.Opts.App,
+		Version:    h.Opts.Deployment,
+		Component:  h.Opts.Component,
+		Node:       h.Opts.Weavelet,
+		TimeMicros: rec.Time.UnixMicro(),
+		Level:      rec.Level.String(),
+		File:       "",
+		Line:       -1,
+		Msg:        rec.Message,
+		Attrs:      appendAttrs(h.Opts.Attrs, attrs),
 	}
-	l.Write(e)
+
+	// Get the file and line information.
+	details := runtime.FuncForPC(rec.PC)
```

Comment:
I was just reading over the docs for `slog`, and I noticed this about `rec.PC` (https://pkg.go.dev/golang.org/x/exp/slog#Record):

	// The only valid use for this value is as an argument to
	// [runtime.CallersFrames]. In particular, it must not be passed to
	// [runtime.FuncForPC].
	PC [uintptr](https://pkg.go.dev/builtin#uintptr)
	
It doesn't say why, but do you think this call is safe?

In `slog`, the `rec.PC` is used with `runtime.CallersFrames`, which I think can get you the same info:

https://github.com/golang/exp/blob/10a507213a2916eb81243cb95bcd7f5e70a771e1/slog/handler.go#L288

https://github.com/golang/exp/blob/10a507213a2916eb81243cb95bcd7f5e70a771e1/slog/record.go#L70-L74

Reply:
Thanks Daniel. Great point.